### ISSUE TITLE
"Too strict maybe" rule is not correct

### ISSUE THREAD

**Issue by athas** [ndmitchell/hlint#881]
`hlint` currently suggests I rewrite this definition:

```
f :: (Show a, Show b) => a -> Maybe b -> String
f a = maybe (show a) show
```

To this:

```
f :: (Show a, Show b) => a -> Maybe b -> String
f a = show . fromMaybe a
```

But this is not actually correct, since the two original `show`s are instantiated at two different types.

**Comment by josephcsible** [comment]
This is the same problem as #87. Short of adding type information to HLint, I don't see a way to fix this without losing a lot of hints that are very useful in most circumstances.

**Comment by athas** [comment]
#87 was fixed by removing the wrong rule, so I think this one should be removed too (or #87 reverted, if the policy has changed).

I think it is a slippery slope for `hlint` to suggest rewrites that are not always correct.

**Comment by josephcsible** [comment]
> I think it is a slippery slope for `hlint` to suggest rewrites that are not always correct.

At least this is always caught at compile time. I'm currently fighting the same fight for another hint that's wrong at runtime in #879.

**Comment by ndmitchell** [comment]
It's sad. Maybe move these hints into a special sometimes correct but not always? Or always correct for monomorphic?

How much do we lose by removing it? How many other hints do you think have a similar issue @josephcsible ?

**Comment by athas** [comment]
What I like about `hlint` is that it is so simple, robust, and predictable (incidentally, this is why I like most of the tools you write @ndmitchell).  I think it's OK to hide rules like this behind an optional flag, but I would be sad if I had to second-guess the default behaviour.

Rules like this one could work for a hypothetical `hlint` alternative that uses the GHC API to actually type-check the program first, but that would be much slower and awkward, for probably not very much gain.  I don't think I would use such an alternative tool.

**Comment by josephcsible** [comment]
I'm working on a list of all the hints we'd lose to this change. However, in doing so, I discovered a much worse bug: #883. 

Anyway, as for what to do about it in general, I think we should add a new note called `unsound`, make sure that `--refactor` always ignores such hints, and print a suitable message on each of them as we output them.

**Comment by josephcsible** [comment]
Looks like we'd lose at least 11 hints:

```haskell
{-# LANGUAGE FunctionalDependencies #-}

import Data.Int (Int8, Int16, Int32)
import Data.Maybe (fromJust)

class IsNextWider a b | a -> b, b -> a

instance IsNextWider Int8 Int16
instance IsNextWider Int16 Int32

andOneWider :: (Num a, Num b, IsNextWider a b) => (a, b)
andOneWider = (0, 0)

oneNumTup :: Num a => (a, a)
oneNumTup = (1, 2)

printInt8 :: Int8 -> IO ()
printInt8 = print

takeInt :: Int -> ()
takeInt _ = ()

takeInteger :: Integer -> ()
takeInteger _ = ()

takeIntAndInteger :: Int -> Integer -> ()
takeIntAndInteger _ _ = ()

badUseSpan :: ([Int], [Integer])
badUseSpan = (takeWhile (< 10) [1..20], dropWhile (< 10) [1..20])

badUseSplitAt :: ([Int], [Integer])
badUseSplitAt = (take 10 [1..20], drop 10 [1..20])

badUseUncurry = takeIntAndInteger (fst oneNumTup) (snd oneNumTup)

-- The hint for this one is ignored by default: - ignore: {lhs: zip (map f x) (map g x), rhs: map (f Control.Arrow.&&& g) x}
badUseAndAndAnd = zip (map takeInt [1..20]) (map takeInteger [1..20])

badRedundantPair :: (Int, Integer)
badRedundantPair = (fst oneNumTup, snd oneNumTup)

badUseBimap = (takeInt (fst oneNumTup), takeInteger (snd oneNumTup))

badUseFirst :: ((), Integer)
badUseFirst = (takeInt (fst oneNumTup), snd oneNumTup)

badUseSecond :: (Int, ())
badUseSecond = (fst oneNumTup, takeInteger (snd oneNumTup))

badTooStrictMaybe1 = maybe (show 123) (show . not)

badTooStrictMaybe2 = maybe (show 123) show (Just False)

-- The hint for this one is in generalise-for-conciseness
badUseBiList :: [Int16]
badUseBiList = [fst andOneWider, snd andOneWider]
```

That code compiles fine. HLint gives 11 meaningfully different hints (one for each `bad*`) for it, though, and applying any of them will cause a compiler error.

**Comment by ndmitchell** [comment]
In all the cases where one expression, which is a value (not a function), is interpreted at different values in the code, I view it as a corner case. We've never actually had a user get bitten by that.

In contrast, whenever its a function, e.g. `show`, a user report pops up sooner or later. Therefore, I'm inclined as a practical matter to disable all those which are involve functions (e.g. too strict maybe), but leave all those that don't (the rest). To get these perfectly right would require type information, which HLint doesn't have.

**Comment by athas** [comment]
That sounds like a good compromise to me.

**Comment by josephcsible** [comment]
Sounds reasonable to me too. One request, though: instead of just disabling/removing those hints, can we keep them, but behind "unsound" or something, so that users who are okay with some false positives can still have them?

**Comment by ndmitchell** [comment]
Sure, although I'd rather have a word for the type of unsoundness. Something like works-apart-from-adhoc-polymorphism?

**Comment by josephcsible** [comment]
How about "requires additional unification"?

**Comment by athas** [comment]
```
--monomorphic
--wrong-but-shorter
--would-work-in-standard-ml
--yolo
```

**Comment by ndmitchell** [comment]
Added them as a monomorphic group. I also edited the manual to include:

> * Sometimes HLint will change the code in a way that causes values to default to different types, which may change the behaviour.
> * HLint assumes duplicate identical expressions within in a single expression are used at the same type.

I think these are the two issues in this ticket and #883.

**Comment by athas** [comment]
Great! I guess that fixes this issue.
